zerver/views/message_fetch.py

```diff
@@ -808,13 +824,37 @@ def get_messages_backend(request: HttpRequest, user_profile: UserProfile,
             MAX_MESSAGES_PER_FETCH,
         ))
 
-    if user_profile.realm.email_address_visibility != Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+    # Web public queries are only allowed for limited types of narrows.
+    web_public_filter = {'operator': 'streams', 'operand': 'web-public', 'negated': False}
+    is_web_public_query = False
+    if narrow is not None and web_public_filter in narrow:
```

Comment:
I'm not a huge an of this `in` check -- it's not super robust.  Let's make a `is_web_public_narrow` function in `zerver/lib/narrow.py` next to `is_web_public_compatible`, perhaps?

Reply:
Okay